src/backend/distributed/planner/query_pushdown_planning.c

```diff
@@ -928,19 +928,91 @@ DeferErrorIfCannotPushdownSubquery(Query *subqueryTree, bool outerMostQueryHasLi
 					  "functions";
 	}
 
-	if (subqueryTree->limitOffset)
+	/*
+	 * Correlated subqueries are effectively functions that are repeatedly called
+	 * for the values of the vars that point to the outer query. We can liberally
+	 * push down SQL features within such a function, as long as co-located join
+	 * checks are applied.
+	 */
+	if (!contain_vars_of_level((Node *) subqueryTree, 1))
 	{
-		preconditionsSatisfied = false;
-		errorDetail = "Offset clause is currently unsupported when a subquery "
-					  "references a column from another query";
-	}
+		if (subqueryTree->limitOffset)
+		{
+			preconditionsSatisfied = false;
+			errorDetail = "Offset clause is currently unsupported when a subquery "
+						  "references a column from another query";
+		}
 
-	/* limit is not supported when SubqueryPushdown is not set */
-	if (subqueryTree->limitCount && !SubqueryPushdown)
-	{
-		preconditionsSatisfied = false;
-		errorDetail = "Limit in subquery is currently unsupported when a "
-					  "subquery references a column from another query";
+		/* limit is not supported when SubqueryPushdown is not set */
```

Comment:
My guess is that with this change we can even consider getting rid of this `LIMIT` hack, where we check ` !SubqueryPushdown`, because as far as I know the customer who relies on this behavior already has correlated subqueries anyway. Otherwise, they query results would be wrong :)

(I think in general `SubqueryPushdown ` is still useful for that customer given that it is a way to skip expensive recursive planning checks).

Though, if we decide to do so, we could double check with them. So, I'm OK to skip this idea, just wanted to note.